enhancements/cluster-scope-secret-volumes/shared-resource-validation.md

```diff
@@ -0,0 +1,407 @@
+the data mounted from the Reference if the Shared Resource is no longer valid.  Default is to remove the data.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Generally, basic field checking of the API Object created or modified when the webhook is called should be sufficient
+for the 'readOnly: true' and 'openshif-*' restriction enforcement.
+
+Any handling of existing Shared Resources whose names start with 'openshift-*' and were created when OCP was at a 
+4.10.x level will be done by the controller, and not the new webhooks.  The controller's existing watches, state around
+running Pods with Shared Resource volume mounts, is better suited to any actions of this ilk.  The controller will use the 
+new "validity" condition, with the name restriction as the reason, if an unapproved use of "openshift-*" is found,
+and will handle mounts and existing data according to the defaults for the 2 new boolean fields.  But let's discuss 
+if we should toggle that behavior.
+
+The webhook will handle the "openshift-*" naming restriction for creates of Shared Resources.
+The webhook will read the "allow list" ConfigMaps on startup (they should not change except on upgrade) and so its validations 
+should not involve any RPC calls.
+
+Existing checks in the controller for the presence of "readOnly: true" check will be kept.  Analogous checks for 
+who creates "openshift-*" Shared Resources will also exist in the controller.  This covers the case when the webhook
+might be down.
+
+Per open questions, discovery of how restricting creation of namespaces/projects starting with "openshift-*" should be
+done.  It is feasible that we augment the ConfigMap based "allow list" with the same mechanism (RBAC based?  Controller based?
+Project endpoint in the aggregated API Server?).
+
+### Risks and Mitigations
+
+Since Shared Resources is Tech Preview 4.10, we can introduce validation controls that "break existing API usage" in later releases than 4.10.
+In particular, any users who were creating SharedSecrets of their own that start with "openshift-".  But we have to 
+have all such restrictions in place before moving from Tech Preview to GA.  The 4.10 blog post does give a warning
+that we might restrict use of Shared Resources names starting with "openshift-*"
+
+And again, the "readOnly: true" validation was already in effect in 4.10, but only enforced at the controller.
+
+That said, even with Tech Preview, release notes are warranted if we go down the path of invalidating of 
+Shared Resources that were "OK" in 4.10.  And user push back is at least conceivable on a case by case basis.
+
+## Design Details
+
+
+### Open Questions [optional]
+
+ 1. Is there a "Pod Spec'able" filter wrt configuring validating admission webhooks, so we catch all the subtypes of Pod?  Or do we have to list each k8s type that has Volumes in the spec?  Do not think there is a "pod spec'able" filter, but let's confirm.
```

Comment:
I wish I knew the answer. IMO, PodSpec can be embedded in CRDs / external API servers and there is no easy way to find them.

Reply:
yeah we'll see if any of the other assignees to this PR know ... thanks

Reply:
Unfortunately I don't think there is a notion of "duck typing" APIs for webhooks - you need to specify the top level APIs to listen to. That would mean we watch things that we ship in OpenShift that reference pod volumes

- Pod itself
- Basic workload types (Deployment, StatefulSet, Job, CronJob, DaemonSet)
- More "inner" workload types (ReplicaSet, ReplicationController)?
- BuildConfig and Build

Reply:
thanks for the additional confirmation @adambkaplan as well ascalling out duck typing, as that is what in part prompted my thinking ... I should have used that phrase as well (just forgot).

Maybe I'll add that phrase at some point, but more than likely I just remove this questions and state the reality.  That said, I should also replace this question with bubbling up the admission webhook vs. plugin question, per your other comments wrt the concerns you and I both voiced on the impact of a misbehaving webhook.

thanks again